### DISCUSSION TITLE
[Question type] Is pattern matching fully explained in documentation?

### DISCUSSION THREAD

**Discussion by trsh** [winkjs/wink-nlp#90] [General]
Document: https://winkjs.org/wink-nlp/custom-entities.html

Statements "Each option is separated by a vertical pipe character as in [NOUN|PROPN]" and "An option may be empty as in the case of the first two sets of options — [|DET] and [|ADJ]" suggest that the symbol "|" can be used as OR operator between 2 options, and if at beginning, the option can be absent.

Are there any more special symbols we can use to achieve certain pattern matching, besides "^" for escaping?

**Comment by sanjayaksaxena** [comment]
Hello @trsh 

Idea to show absence right in the beginning is more of a best practice — basically to avoid any inadvertent miss.

There is only a single character — `^` available for escaping; if you wish to detect it in the pattern then please use `^^`.

We can help further if you explain the use-case.

Best,
Sanjaya

**Comment by trsh** [comment]
@sanjayaksaxena thank you for your reply. 3 questions.

1) In my case the patterns can be tricky, something like a email, but different problem that would require a regex. But I see its not supported for custom entities? What would you suggest in such case?

2) Also a more simple case is Nouns that can be plural or singular, but should match in both cases. What would you recommend?

**Reply by trsh** [reply]
3. And how about matching entities that are very similar? Use case would be a misspelled word in text. Like someone wrote "bonana" instead of "banana".

**Reply by sanjayaksaxena** [reply]
1. Regex is not supported now. We are currently evaluating the possibility to support regex inside a pattern.

3. Current version only supports `value` or `normal` in the configuration. We will be adding `lemma` as an option in the upcoming version.
4. For misspelled words suggest that use options i.e. `[banana|bonana]`.

**Reply by trsh** [reply]
1. Great.
2. How does `value` and `normal` connect to my question? I must be missing something here.
3. This is not an option, because I don't know input. It can be misspelled in lots of ways. That's why I ask about similarity feature.

**Reply by sanjayaksaxena** [reply]
1. Please refer to **config** section at [https://winkjs.org/wink-nlp/learn-custom-entities.html](https://winkjs.org/wink-nlp/learn-custom-entities.html).
2. Similarity inside pattern is not available. You can leverage [wink-distance](https://winkjs.org/wink-distance/) — it provides distances/similarities like [levenshtein](https://winkjs.org/wink-distance/string.html#.levenshtein), [jaccard](https://winkjs.org/wink-distance/set.html#.jaccard) and more.
